runtime/compiler/ilgen/IlGenerator.cpp

```diff
@@ -564,7 +564,8 @@ TR_J9ByteCodeIlGenerator::genILFromByteCodes()
 
       if (currNode->getOpCodeValue() == TR::checkcast
           && currNode->getSecondChild()->getOpCodeValue() == TR::loadaddr
-          && currNode->getSecondChild()->getSymbolReference()->isUnresolved())
+          && currNode->getSecondChild()->getSymbolReference()->isUnresolved()
+          && !TR::Compiler->cls.isClassRefValueType(method()->classOfMethod(), currNode->getSecondChild()->getSymbolReference()->getCPIndex(), comp()))
```

Comment:
it seems wweird to say the the symbolReference isUnresovled and then as a question about it. A comment to explain this is probably warranted...

Reply:
Added comments as suggested.

Reply:
Yi @cathyzhyi, I was looking at your adjusted comment.  Just so I'm clear on this, the class specified on `CHECKCAST` is required to be resolved only if the object is non-null and the class is not a value type.  If the class is a value type, it will never be resolved by the `CHECKCAST`.  Right?

Reply:
Hey Henry, @hzongaro the class reference can still be resolved by `CHECKCAST` if the object is not-null and is value type. The value type class is excluded from expansion because the resolve check is already inserted when calling `loadClassObject`. The expansion for unresolved case is needed for reference type because we want to avoid resolving the class when the object is null so can't just insert the resolve check in the mainline path for all cases as we do for value type.

Reply:
Thanks, Yi @cathyzhyi!  Now I see.

I've followed through the logic more carefully now, and I'm wondering whether the name of `unresolvedCheckcastTops` should be changed to make it clear that not all the unresolved CHECKCASTs are included - only the ones that need to be expanded.  Maybe something like `unexpandedUnresolvedCheckcastTops`?  Or `unresolvedCheckcastTopsNeedingNullGuard`?  (I know those are kind of long names - I'm really good at names that are way too long.  :-)

Reply:
Thanks for the suggestion! Changed the variable name to `unresolvedCheckcastTopsNeedingNullGuard`